src/lib.rs

```diff
@@ -4,57 +4,88 @@ extern crate std;
 use alloc::format;
 use alloc::string::{String, ToString};
 use std::io::{BufReader, Read};
+use sha2::{Sha256, Digest};
 
-pub fn generate_sha1_git_oid(x: &[u8]) -> String {
-    let prefix = format!("blob {}\0", x.len());
-    let mut hasher = sha1_smol::Sha1::new();
+use base64ct::{Base64, Encoding};
 
-    hasher.update(prefix.as_bytes());
-    hasher.update(x);
+enum HashAlgorithm {
+    SHA1,
+    SHA256
+}
 
-    hasher.digest().to_string()
+struct GitOid {
+    hash_algorithm: HashAlgorithm,
 }
 
-pub fn generate_sha1_git_oid_from_buffer<R>(
-    mut reader: BufReader<R>,
-    expected_length: usize,
-) -> String
-where
-    BufReader<R>: std::io::Read,
-{
-    let prefix = format!("blob {}\0", expected_length);
-    let mut hasher = sha1_smol::Sha1::new();
-
-    hasher.update(prefix.as_bytes());
-
-    let mut buf = [0; 4096]; // linux default page size is 4096
-    let mut amount_read = 0;
-    loop {
-        let y = reader.read(&mut buf);
-        match y {
-            Ok(0) => {
-                break;
-            }
-            Ok(size) => {
-                hasher.update(&buf[..size]);
-                amount_read = amount_read + size;
-            }
-            Err(_) => {
-                break;
+impl GitOid {
+    pub fn generate_git_oid(&self, x: &[u8]) -> String {
+        let prefix = format!("blob {}\0", x.len());
+
+        match self.hash_algorithm {
+            HashAlgorithm::SHA1 => {
+                let mut hasher = sha1_smol::Sha1::new();
+
+                hasher.update(prefix.as_bytes());
+                hasher.update(x);
+
+                hasher.digest().to_string()
+            },
+            HashAlgorithm::SHA256 => {
+                let mut hasher = Sha256::new();
+                hasher.update(x);
+
+                let hash = hasher.finalize();
+                println!("Binary hash: {:?}", hash);
+
+                let hash_string = Base64::encode_string(&hash);
```

Comment:
I found examples of using both base 16 and base 64 to encode the hash into a string. I'm not hugely attached to either of them, are there any opinions on which would be better to use?

Reply:
@nellshamrell What about good old hex ? :)

Reply:
I'd be ok with merging it and doing a follow-up PR. The enum pattern is excellent here and solves all the issues I had with the type system.

I also will move this from sha1_smol to sha1 once we get this in since sha1 and sha2 are based on the same platform.

Reply:
Cool! I'm going to add one more thing refactoring wise then mark this as ready for review :)